site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -0,0 +1,200 @@
+But because it's a standard, this means performance and features can vary widely between operating systems, graphics driver, and the physical graphics card.
+As you might have guessed, this leads to some [OS specific bugs that are very hard to track down](https://github.com/citra-emu/citra/issues/2416).
+In the linked issue, only on Mac OSX, Citra would leak memory from the hardware renderer.
+We traced it back to how textures were juggled between the 3DS memory and the host GPU, but we don't have many developers that use Mac, so we never did find the root cause.
+For a little bit of good news, **this is fixed in the latest nightly**, but only because the entire texture handling code was rewritten!
+
+## Moving Forward with the Hardware Renderer: Cleaning up Texture Forwarding
+
+Despite the issues mentioned above, OpenGL has been a fair choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making improving the renderer.
+Their first major contribution was a massive, [complete rewrite](https://github.com/citra-emu/citra/pull/3281) of the [texture forwarding support](https://citra-emu.org/entry/texture-forwarding-brings-hd-output-to-citra/) that was added back in 2016.
+The new texture forwarding code increases the speed of many games, and fixes upscaled rendering in some other games as well.
+
+{{< figure src="/images/entry/improvements-to-hardware-renderer/metroid_before.png"
+    title="Metroid: Samus Returns heavily used a PICA200 feature that had a comment `// TODO: implement this` in the hardware renderer" >}}
+{{< figure src="/images/entry/improvements-to-hardware-renderer/metroid_after.png"
+    title="Writing that missing method fixed upscaling while making the game much faster" >}}
+
+Whenever a texture is used in the hardware renderer, the hardware renderer will try to use a copy of the texture already in the GPU memory, but if that fails, it has to reload the texture from the emulated 3DS memory.
+This is called a texture upload, and it's slow for a good reason.
+The communication between CPU and GPU is optimized for large amounts of data transferred, but as a tradeoff, it's not very fast.
+This works great for PC games, where you know all the textures you want to upload ahead of time and can send them in one large batch, but ends up hurting performance for Citra since we can't know in advance when the emulated game will do something that requires a texture upload.
+
+The texture forwarding rewrite increases the speed of many games by adding in new checks to avoid this costly syncronization of textures between emulated 3DS memory and the host GPU memory.
+Additionally, the new texture forwarding can avoid even more texture uploads by copying the data from an existing and compatible locations.
+As an extension of this feature, phantom went the extra mile and fixed Pokémon outlines as well!
```

Comment:
nit: IMO, all username references should be linked. Makes the sentences read less weird without capitalization